File: 프로그래머스/lv2/42888. 오픈채팅방/오픈채팅방.py
def solution(record):
    arr = dict()
    answer = []
    
    for i in record:
        i = i.split()
        if i[0] == "Enter":
            answer.append([i[1], "님이 들어왔습니다."])
            arr[i[1]] = i[2]
        elif i[0] == "Leave":
            answer.append([i[1], "님이 나갔습니다."])
        else:
            arr[i[1]] = i[2]
    
    answer = list(map(lambda x : arr[x[0]] + x[1], answer))
    return answer
    

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;
class Solution {
    static HashMap<String, String> name = new HashMap<>();
    static List<String[]> orders = new ArrayList<>();
    public String[] solution(String[] record) {
        for(String r : record){
            String[] tmp = r.split(" ");
            if(tmp[0].equals("Leave")){
                orders.add(new String[]{tmp[1], "님이 나갔습니다."});
            }
            else if(tmp[0].equals("Change")){
                name.put(tmp[1], tmp[2]);
            }
            else{
                if(name.get(tmp[1]) == null)
                    name.put(tmp[1], tmp[2]);    
                else
                    name.put(tmp[1], tmp[2]);
                orders.add(new String[]{tmp[1], "님이 들어왔습니다."});
            }
        }
        
        String[] answer = new String[orders.size()];
        int i = 0;
        for(String[] order: orders){
            answer[i++] = name.get(order[0]) + order[1];
        }
        return answer;
    }
}